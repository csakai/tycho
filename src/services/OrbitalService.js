import * as THREE from 'three';
import Math2 from './Math2';
import Scale from '../utils/Scale';

export default class OrbitalService {
  
  /**
   * Calculates inclination at a right angle if parity matches.
   *
   * @param {Number} inclination - angle of inclination, in degrees
   * @param {Boolean} [odd = false] - parity
   * @returns {Number} calculated inclination
   */
  static getInclination = (inclination, odd = false) => {
    if (odd) {
      return inclination;
    }
    return inclination - 90;
  }

  /**
   * Calculates the Eulerian vector of the ecliptic group
   *
   * @param {Object} props - OrbitalContainer props
   * @param {Number} props.inclination - angle of inclination, in degrees
   * @param {Number} props.longitudeOfAscendingNode - longitude of ascending node, in degrees
   * @param {Boolean} props.odd - orbital parity
   * @returns {THREE.Euler} Eulerian vector
   */
  static getEclipticGroupRotation = ({inclination, longitudeOfAscendingNode, odd}) => {
    return OrbitalService.toEuler({
      x: OrbitalService.getInclination(inclination, odd),
      z: longitudeOfAscendingNode
    });
  }

  /**`
   * Calculates the Eulerian vector of the orbital group
   *
   * @param {Object} props - OrbitalContainer props
   * @param {Number} props.argumentOfPeriapsis - argument of periapsis, in degrees
   * @returns {THREE.Euler} Eulerian vector
   */
  static getOrbitalGroupRotation = ({argumentOfPeriapsis}) => {
    return OrbitalService.toEuler({
      z: argumentOfPeriapsis
    });
  }

  /**
   * Calculates 3D rotation of the spherical orbital body.
   *
   * @param {Object} props - OrbitalContainer props
   * @param {Number} props.axialTilt - axis tilt
   * @param {Number} props.arcRotate - orbital rotation, in arcseconds
   * @param {Number} props.time - current time, in seconds
   * @returns {THREE.Euler} Eulerian vector
   */
  static getBodyRotation = ({axialTilt, arcRotate, time}) => {
    return OrbitalService.toEuler({
      x: Math.abs(90 - axialTilt),
      y: Math2.arcSecToDeg(time, arcRotate),
    });
  }

  /**
   * Returns the present position vector w.r.t. the elliptical plane.
   *
   * @param {Object} props - OrbitalContainer props
   * @param {Object} props.periapses - Orbital periapses (TODO: change to number, as orbitalPeriod)
   * @param {Number} props.time - current time, in seconds
   * @param {Ellipse} ellipse - instance of ellipse
   * @returns {THREE.Vector3} projected body position
   */
  static getBodyPosition = ({periapses, time}, ellipse) => {
    return ellipse.getPosition(time, periapses);
  }

  /**
   * Scales the body radius by the global scaling factor.
   *
   * @param {Object} props - OrbitalContainer props
   * @param {Number} props.radius - body radius, in kilometers
   * @param {Number} props.scale=1 - scalar multiple
   * @returns {Number} scaled body radius
   */
  static getBodyRadius = ({radius, scale}) => {
    return Scale(radius, scale);
  }

  /**
   * Translate 3D space coordinates to 2D screen ones.
   *
   * @param {THREE.Vector3} position - current position in ellipse
   * @param {THREE.Camera} camera - active renderer camera
   * @returns {Object} coordinates with `top` and `left` values
   */
  static translateWorldToScreen = (position, camera) => {
    const width = window.innerWidth;
    const height = window.innerHeight;
    const matrix = new THREE.Matrix4();

    if (position && camera) {
      const pos = position.clone();
      
      matrix.multiplyMatrices(camera.projectionMatrix, camera.matrixWorldInverse);
      pos.applyMatrix4(matrix);
      
      const left = (1 + pos.x) * width / 2;
      const top = (1 - pos.y) * height / 2;
      
      return {top, left};
    }
    return null;
  }

  /**
   * Returns the vector w.r.t. <0> of the given mesh.
   *
   * @param {THREE.Object3D} mesh - object3d to get vector position of
   * @returns {THREE.Vector3} world position
   */
  static getWorldPosition = (mesh) => {
    if (mesh) {
      const vect = new THREE.Vector3();
      const matrix = mesh
        ._reactInternalInstance
        ._threeObject
        .matrixWorld;

      vect.setFromMatrixPosition(matrix);

      return vect;
    }
    return null;
  }

  /**
   * Converts a standard 3D vector to an Eulerian one.
   *
   * @param {Object} coords - coordinates
   * @param {Number} [coords.x = 0] - x coordinate
   * @param {Number} [coords.y = 0] - y coordinate
   * @param {Number} [coords.z = 0] - z coordinate
   * @returns {THREE.Euler} Eulerian vector
   */
  static toEuler = ({x, y, z}) => {
    const toRad = val => val ? Math2.toRadians(val) : 0;
    return new THREE.Euler(toRad(x), toRad(y), toRad(z));
  }
}

