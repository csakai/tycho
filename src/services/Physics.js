export default class Physics {

  static MAX_ECCENTRICITY = 0.8

  static KM_TO_AU = 6.68459e-9;

  static GRAVITATIONAL_CONSTANT = 6.67408e-11;

  /**
   * Calclates the mean anomaly in degrees.
   * Mean anomaly is the angular distance from 
   * the focus of an orbit to the present position.
   *
   * @param {Number} time - time passed, in seconds
   * @param {Number} period - period of rotation
   * @returns {Number} calculated mean anomaly
   */
  static meanAnomaly(t, P) {
    return t / P % 1 * 360;
  }

  /**
   * Calculates the eccentric anomaly.
   * Eccentric anomaly is the angular distance from 
   * the focus of an orbit to the present position.
   *
   * @param {Number} ecc - eccentricity of ellipse
   * @param {Number} time - present time, in seconds
   * @param {Number} periapses - periapses of revolution
   * @returns {Number} calculated eccentric anomaly
   */
  static eccentricAnomaly(ecc, time, periapses) {
    const timePassed = (time - periapses.last);
    const period = (periapses.next - periapses.last);

    let meanAnomaly = this.meanAnomaly(timePassed, period) / 360;
    let E, F;
    
    meanAnomaly = 2 * Math.PI * (meanAnomaly - Math.floor(meanAnomaly));
    E = ecc < this.MAX_ECCENTRICITY ? meanAnomaly : Math.PI;
    F = E - ecc * Math.sin(meanAnomaly) - meanAnomaly;
    
    // numerical approximation for Kepler's second law
    for(let i = 0; i < 10; i++) {
      E = E - F / (1 - ecc * Math.cos(E));
      F = E - ecc * Math.sin(E) - meanAnomaly;
    }
    return E;
  }
  
  /**
   * Returns the projected orbital angle from eccentricity and eccentric anomaly.
   *
   * @param {Number} ecc - eccentricity
   * @param {Number} E - eccentric anomaly, in degrees
   * @returns {Number} projected orbital angle
   */
  static getTheta(ecc, E) {
    const halfPi = Math.PI / 180;
    const min = Math.sqrt(1 - Math.pow(ecc, 2));
    const theta = Math.atan2(
      (min * Math.sin(E)),
      (Math.cos(E) - ecc)
    ) / halfPi;
    
    if (theta < 0) {
      return 360 + theta;
    }
    return theta;
  }

  /**
   * Converts an angle theta, in degrees, to percentage.
   *
   * @param {Number} theta - angle in degrees
   * @returns {Number} percentage of ellipse completed
   */
  static thetaToPercent(theta) {
    const percent = theta / 360;

    if(percent > 1 || isNaN(percent)) {
      return 0;
    }
    return percent;
  }

  /**
   * Returns present percentage of ellipse travelled.
   *
   * @param {Number} ecc - eccentricity of ellipse
   * @param {Number} time - present time, in seconds
   * @param {Object} periapses - {next, last} period, in seconds
   * @returns {Number} percetnage of ellipse travelled
   */
  static ellipticPercent(ecc, time, periapses) {
    let E = this.eccentricAnomaly(ecc, time, periapses);
    let theta = this.getTheta(ecc, E);
    
    return this.thetaToPercent(theta);
  }
  
  /**
   * Calculates a mass' specific orbital energy constant at present vector.
   * Utilizes the calculation derived from Kepler's First Law.
   *
   * @param {Number} mass - mass of the orbital body, in kilograms
   * @param {Number} magnitude - magnitude of vector to origin, in kilometers
   * @param {Number} semimajor - size of semimajor axis, in kilometers
   * @returns {Number} orbital energy constant at present vector, in km/s
   */
  static orbitalEnergyConservation(centralMass, magnitude, semimajor) {
    const a = semimajor * 1000; // km to m
    const r = magnitude * 1000; // km to m
    const GM = Physics.GRAVITATIONAL_CONSTANT * centralMass; // m^3/s^2
    const speed = Math.sqrt(GM * ((2 / r) - (1 / a))); // m/s
  
    return speed / 1000; // m to km
  }
  
  /**
   * Converts given kilometers to astronomical units.
   *
   * @param {Number} x - kilometers
   * @param {Number} scale = 1 - scaling factor
   * @returns {Number} result in astronomical units
   */
  static toAU(x, scale = 1) {
    return x * scale * this.KM_TO_AU;
  }
}
