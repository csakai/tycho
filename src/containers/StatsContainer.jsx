import React from 'react';
import moment from 'moment';
import { connect } from 'react-redux';
import ReduxService from '../services/ReduxService';
import OrbitalService from '../services/OrbitalService';
import Constants from '../constants';
import Stats from '../components/Stats';

export class StatsContainer extends React.Component {

    componentWillMount = () => {
        this.setState({});
    }

    componentWillReceiveProps = (nextProps) => {
        const { targetId, time } = nextProps;

        if (this.props.targetId !== targetId || this.props.time !== time) {
            this.updateOrbitalStats(targetId, time);
        }
    }

    /**
     * Updates orbital information.
     *
     * @param {String} targetId - name of active target
     * @param {Number} time - current UNIX time
     */
    updateOrbitalStats = (targetId, time) => {
        const { orbitalData } = this.props;
        const target = OrbitalService.getTargetByName(orbitalData, targetId);

        if (target) {
            const { name, description } = target;
            this.setState({
                name,
                description,
                ...OrbitalService.getOrbitalStats(target, time)
            });
        }
    }

    /**
     * Returns the current UNIX time prop in UI-friendly format.
     *
     * @returns {String} UI-friendly time
     */
    getTime = () => {
        return moment(this.props.time * 1000)
            .format(Constants.UI.UX_DATE_FORMAT);
    }

    render() {
        return (
            <Stats
                description={this.state.description}
                velocity={this.state.velocity}
                magnitude={this.state.magnitude}
                trueAnomaly={this.state.trueAnomaly}
                pageText={this.props.pageText}
                time={this.getTime()}
            />
        );
    }
}

export default connect(
    ReduxService.mapStateToProps(
        'label.targetId',
        'data.orbitalData',
        'data.pageText',
        'animation.time'
    ),
    null
)(StatsContainer);
