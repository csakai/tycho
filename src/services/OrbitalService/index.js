import OrbitalService from './OrbitalService';

export default OrbitalService;
