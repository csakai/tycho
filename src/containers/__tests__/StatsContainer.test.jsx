import React from 'react';
import toJson from 'enzyme-to-json';
import {shallow} from 'enzyme';
import {StatsContainer} from '../StatsContainer';
import OrbitalService from '../../services/OrbitalService';
import Physics from '../../services/Physics';
import orbitalData from './__fixtures__/orbitals.json';

describe('Modal Container', () => {
  let component, statsContainer;

  beforeEach(() => {
    component = shallow(
      <StatsContainer
        orbitalData={orbitalData}
        targetId="dummyPlanet"
      />
    );
    statsContainer = component.instance();
  });

  describe('componentWillReceiveProps()', () => {
    beforeEach(() => {
      OrbitalService.getTargetByName = () => orbitalData[0];
    });

    it('should call updateTargetParams if the targetId has changed', () => {
      const spy = jest.spyOn(statsContainer, 'updateTargetParams');
      const targetId = 'Mars';
      const nextProps = {targetId};

      statsContainer.props = {targetId: 'Earth'};
      statsContainer.componentWillReceiveProps(nextProps);

      expect(spy).toHaveBeenCalled();
      expect(spy).toHaveBeenCalledWith(targetId);
    });

    it('should not call updateTargetParams if the targetId remains the same', () => {
      const spy = jest.spyOn(statsContainer, 'updateTargetParams');
      const targetId = 'Earth';
      const nextProps = {targetId};

      statsContainer.props = {targetId};
      statsContainer.componentWillReceiveProps(nextProps);

      expect(spy).not.toHaveBeenCalled();
    });
  });

  describe('updateTargetParams()', () => {
    it('should update the state with the current target\'s params', () => {
      const data = orbitalData[0];

      OrbitalService.getTargetByName = () => data;
      statsContainer.updateTargetParams();

      expect(component.state('name')).toEqual(data.name);
      expect(component.state('description')).toEqual(data.description);
      expect(component.state('GM')).toEqual(data.GM);
      expect(component.state('semimajor')).toEqual(data.semimajor);
    });
  });

  describe('updateOrbitalStats()', () => {
    it('should update the component state with the velocity and magnitude', () => {
      OrbitalService.getDistanceToSun = () => 5;
      Physics.orbitalEnergyConservation = () => 4;
      statsContainer.updateOrbitalStats();

      expect(component.state('magnitude')).toEqual('5.000');
      expect(component.state('velocity')).toEqual('4.000');
    });
  });

  describe('render()', () => {
    it('should render the modal successfully', () => {
      expect(toJson(component)).toMatchSnapshot();
    });
  });
});
