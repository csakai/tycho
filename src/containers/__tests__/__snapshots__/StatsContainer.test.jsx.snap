// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Stats Container render() should render the modal successfully 1`] = `
<Stats
  time="Invalid date"
/>
`;
